packages/auth/lib/AuthProvider.ts

```diff
@@ -0,0 +1,140 @@
+      communicationAddress: null,
+      createdAt: randomTime,
+      userAuthenticationId: uuid.v4(),
+      userId: uuid.v4()
+    };
+
+    return {
+      authFactor,
+      passwordData
+    };
+  }
+
+  // tslint:disable-next-line:prettier
+  public async create(password: string, communicationAddress: string = null, isProofed: boolean = false, providerMeta: any = {}): Promise<string> {
+    const sodiumConfig = createConfig(this.authConfig.sodium);
+
+    const providerSignature = this.signHelper.getProviderSignature(this.authConfig.secrets.authProviderHashSignature, this.providerName, "");
+
+    const passwordData: IPasswordData = await newHash(password + providerSignature, sodiumConfig);
+
+    const authFactorCreation: IAuthFactorCreation = {
+      provider: this.providerName,
+      communicationAddress,
+      isProofed,
+      meta: JSON.stringify({ sodiumMeta: passwordData.meta, providerMeta }),
+      hash: passwordData.hash
+    };
+
+    return this.authConnector.createAuthFactorCreationToken(authFactorCreation);
+  }
+
+  public async proof(userIdentifier: string, getPassword: (authFactor: IAuthFactorForProof) => Promise<string>): Promise<IProofResponse> {
+    const provider = this.providerName;
+    let authFactor: IAuthFactorForProof;
+    let passwordData: IPasswordData;
+    let isFake = false;
+
+    try {
+      authFactor = await this.authConnector.getAuthFactorForProof(userIdentifier, provider);
+
+      const meta = JSON.parse(authFactor.meta);
+
+      const password = await getPassword(authFactor);
+
+      const providerSignature = this.signHelper.getProviderSignature(
+        this.authConfig.secrets.authProviderHashSignature,
+        meta.isOldPassword === true ? "local" : provider,
+        meta.isOldPassword === true ? authFactor.userId : ""
+      );
+
+      passwordData = await hashByMeta(password + providerSignature, meta.sodiumMeta);
+    } catch (err) {
+      // TODO: Log this
+      const randomAuthFactor = await this.createRandomAuthFactor();
+      authFactor = randomAuthFactor.authFactor;
+      passwordData = randomAuthFactor.passwordData;
+      isFake = true;
+
+      try {
+        await getPassword(authFactor);
+      } catch (err) {
+        /* Ignore Error: This is only called to spent exactly the same time in case of an error as in case of validation. */
+      }
+    }
+
+    const authFactorProof: IAuthFactorProof = {
+      id: authFactor.id,
+      hash: passwordData.hash,
+      provider,
+      maxAgeInSeconds: this.authFactorProofTokenMaxAgeInSeconds
+    };
+
+    return {
+      authFactorProofToken: this.authConnector.createAuthFactorProofToken(authFactorProof),
+      isFake
+    };
+  }
+
+  /* For anyone who wants to refactor something: 
+    You may want to use this function inside the proof method. 
+    However, if you do the time spent to create a fake auth-factor would be bigger than validating a correct auth-factor.
```

Comment:
Good comment. However, this sounds a bit confusing to me. Maybe:
```
Do not use this function inside the proof method.
If you do, the time for creating a fake auth-factor would be bigger than the time for validating a correct auth-factor.
```

Reply:
:+1: Changed in 1077cd16837c47280034e7875c14e94a63e6be30